subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java

```diff
@@ -231,29 +227,33 @@ public void putBytes(byte[] bytes, int off, int len) {
             digest.update(bytes, off, len);
         }
 
-        @Override
-        public HashCode hash() {
-            done = true;
-            byte[] bytes = digest.digest();
-            return HashCode.fromBytesNoCopy(bytes);
-        }
-
-        private void update(int length) {
-            checkNotDone();
-            digest.update(buffer.array(), 0, length);
-            BufferCaster.cast(buffer).clear();
-        }
-
         @Override
         public void putInt(int value) {
-            buffer.putInt(value);
-            update(4);
+            byte[] bytes = new byte[]{
+                (byte) value,
+                (byte) (value >> 8),
+                (byte) (value >> 16),
+                (byte) (value >> 24)
+            };
+            // TODO change to bigendian Ints.toByteArray(value);
+            // TODO: apply to org.gradle.internal.hash.HashCode.fromInt, too
+            putBytes(bytes);
         }
 
         @Override
         public void putLong(long value) {
-            buffer.putLong(value);
-            update(8);
+            byte[] bytes = new byte[]{
+                (byte) value,
+                (byte) (value >> 8),
+                (byte) (value >> 16),
+                (byte) (value >> 24),
+                (byte) (value >> 32),
+                (byte) (value >> 40),
+                (byte) (value >> 48),
+                (byte) (value >> 56)
+            };
```

Comment:
Originally it was done to enable reuse (i.e. less stateful). When no ints or longs are created, this results in less garbage. When more are (like the sizes before byte arrays), this can indeed result in more garbage, unless escape analysis detects it.
Would you like me to check if it does?
 
This also seemed less error prone to me (e.g. like leftover garbage after a `long` conversion that can be left there when reading an `int`).
My measurements show that it has the same speed.

Would you like me to revert this part?

Reply:
I think we should revert it. Most hashers will include some ints at least (for length prefixes), so I don't expect this to save anything (unless escape analysis or some other dark JVM magic picks this up, but I don't expect that to be the case). If we could identify that the "no ints or longs" case is widespread, we could make the allocation of the buffer be lazy, but again, or use a pool of buffers etc.

That said, I don't think we should go any further in this direction now.

Reply:
Thanks, reverted.